cloudinit/sources/helpers/vultr.py

```diff
@@ -0,0 +1,293 @@
+
+    v1_json = json.loads(v1)
+    metadata = v1_json
+
+    # This comes through as a string but is JSON, make a dict
+    metadata['vendor-config'] = json.loads(metadata['vendor-config'])
+
+    return json.dumps(metadata)
+
+
+def get_cached_metadata(args):
+    return json.loads(get_metadata(json.dumps(args)))
+
+
+# Read the system information from SMBIOS
+def get_sysinfo():
+    return {
+        'manufacturer': dmi.read_dmi_data("system-manufacturer"),
+        'subid': dmi.read_dmi_data("system-serial-number")
+    }
+
+
+# Assumes is Vultr is already checked
+def is_baremetal():
+    if get_sysinfo()['manufacturer'] != "Vultr":
+        return True
+    return False
+
+
+# Confirm is Vultr
+def is_vultr():
+    # VC2, VDC, and HFC use DMI
+    sysinfo = get_sysinfo()
+
+    if sysinfo['manufacturer'] == "Vultr":
+        return True
+
+    # Baremetal requires a kernel parameter
+    if "vultr" in util.get_cmdline():
+        return True
+
+    return False
+
+
+def convert_to_base64(string):
+    string_bytes = string.encode('ascii')
+    b64_bytes = base64.b64encode(string_bytes)
+    return b64_bytes.decode('ascii')
+
+
+# Read Metadata endpoint
+def read_metadata(params):
+    response = url_helper.readurl(params['url'],
+                                  timeout=params['timeout'],
+                                  retries=params['retries'],
+                                  headers={'Metadata-Token': 'vultr'},
+                                  sec_between=params['wait'])
+
+    if not response.ok():
+        raise RuntimeError("Failed to connect to %s: Code: %s" %
+                           params['url'], response.code)
+
+    return response.contents.decode()
+
+
+# Get Metadata by flag
+def fetch_metadata(params):
+    req = dict(params)
+    req['url'] = "%s/v1.json" % params['url']
+
+    return read_metadata(req)
+
+
+# Wrapped for caching
+@lru_cache()
+def get_interface_map():
+    return net.get_interfaces_by_mac()
+
+
+# Convert macs to nics
+def get_interface_name(mac):
+    macs_to_nic = get_interface_map()
+
+    if mac not in macs_to_nic:
+        return None
+
+    return macs_to_nic.get(mac)
+
+
+# Generate network configs
+def generate_network_config(config):
+    md = get_cached_metadata(config)
+
+    network = {
+        "version": 1,
+        "config": [
+            {
+                "type": "nameserver",
+                "address": [
+                    "108.61.10.10"
+                ]
+            }
+        ]
+    }
+
+    # Prepare interface 0, public
+    if len(md['interfaces']) > 0:
+        network['config'].append(generate_public_network_interface(md))
+
+    # Prepare interface 1, private
+    if len(md['interfaces']) > 1:
+        network['config'].append(generate_private_network_interface(md))
+
+    return network
+
+
+# Input Metadata and generate public network config part
+def generate_public_network_interface(md):
+    interface_name = get_interface_name(md['interfaces'][0]['mac'])
```

Comment:
Am I right that this only ever uses md['interfaces'] ?  If so, you can have the one caller pass that in.